libvirt/resource_libvirt_domain.go

```diff
@@ -825,72 +848,102 @@ func resourceLibvirtDomainDelete(d *schema.ResourceData, meta interface{}) error
+			} else if !found && !ignore {
+				// re-add the interface and deal with it later
+				meta.(*DomainMeta).ifaces <- iface
+			}
 		}
+
+		return meta, state, nil
 	}
 }
 
-func waitForNetworkAddresses(ifaces []defNetworkInterface, domain libvirt.VirDomain) error {
-	log.Printf("[DEBUG] waiting for network addresses.\n")
-	// wait for network interfaces with 'wait_for_lease' to get an address
-	for _, iface := range ifaces {
-		if !iface.waitForLease {
-			continue
-		}
-
-		mac := strings.ToUpper(iface.Mac.Address)
-		if mac == "" {
-			log.Printf("[DEBUG] Can't wait without a mac address.\n")
-			// we can't get the ip without a mac address
-			continue
-		}
+func hasNetworkAddress(iface defNetworkInterface,
+	domain libvirt.VirDomain) (found bool, ignore bool, err error) {
 
-		// loop until address appear, with timeout
-		start := time.Now()
+	if !iface.waitForLease {
+		return false, true, nil
+	}
 
-	waitLoop:
-		for {
-			log.Printf("[DEBUG] waiting for network address for interface with hwaddr: '%s'\n", iface.Mac.Address)
-			ifacesWithAddr, err := getDomainInterfaces(&domain)
-			if err != nil {
-				return fmt.Errorf("Error retrieving interface addresses: %s", err)
-			}
-			log.Printf("[DEBUG] ifaces with addresses: %+v\n", ifacesWithAddr)
+	mac := strings.ToUpper(iface.Mac.Address)
+	if mac == "" {
+		log.Printf("[DEBUG] Can't wait without a mac address.\n")
+		// we can't get the ip without a mac address
+		return false, true, nil
+	}
 
-			for _, ifaceWithAddr := range ifacesWithAddr {
-				// found
-				if mac == strings.ToUpper(ifaceWithAddr.Hwaddr) {
-					break waitLoop
-				}
-			}
+	log.Printf("[DEBUG] waiting for network address for interface with hwaddr: '%s'\n", iface.Mac.Address)
+	ifacesWithAddr, err := getDomainInterfaces(&domain)
+	if err != nil {
+		return false, false, fmt.Errorf("Error retrieving interface addresses: %s", err)
+	}
+	log.Printf("[DEBUG] ifaces with addresses: %+v\n", ifacesWithAddr)
 
-			time.Sleep(1 * time.Second)
-			if time.Since(start) > 5*time.Minute {
-				return fmt.Errorf("Timeout waiting for interface addresses")
-			}
+	for _, ifaceWithAddr := range ifacesWithAddr {
+		// found
+		if mac == strings.ToUpper(ifaceWithAddr.Hwaddr) {
+			return true, false, nil
 		}
 	}
 
-	return nil
+	return false, false, nil
+}
+
+func getDomainState(domain libvirt.VirDomain) (string, error) {
+	state, err := domain.GetState()
+	if err != nil {
+		return "", err
+	}
+
+	var stateStr string
+
+	switch state[0] {
+	case libvirt.VIR_DOMAIN_NOSTATE:
+		stateStr = "nostate"
+	case libvirt.VIR_DOMAIN_RUNNING:
+		stateStr = "running"
+	case libvirt.VIR_DOMAIN_BLOCKED:
+		stateStr = "blocked"
+	case libvirt.VIR_DOMAIN_PAUSED:
+		stateStr = "paused"
+	case libvirt.VIR_DOMAIN_SHUTDOWN:
+		stateStr = "shutdown"
+	case libvirt.VIR_DOMAIN_CRASHED:
+		stateStr = "crashed"
+	case libvirt.VIR_DOMAIN_PMSUSPENDED:
+		stateStr = "pmsuspended"
+	case libvirt.VIR_DOMAIN_SHUTOFF:
+		stateStr = "shutoff"
+	}
```

Comment:
I would prefer to have also a default state to handle the ones that might be added in the future, something like:

```golang
default:
  stateStr = fmt.Sprintf("unknown - %v", state[0])
```